Sign In story PR

aabdoo23 opened the pull request:
#### PR Classification
New feature: Introduces repository pattern, authentication services, and database context setup for the SchedsForums application.

#### PR Summary
This PR adds new controllers, services, repositories, and database context configurations to support user authentication and student management.
- **LoginController.cs** and **StudentController.cs**: Added new endpoints for login and student creation using MediatR.
- **Program.cs**: Registered services for database context, repositories, MediatR, and JWT authentication.
- **SchedsForumsDbContext.cs**: Defined DbContext with DbSet properties for users, admins, students, and moderators.
- **CreateStudentCommand.cs** and **LoginCommand.cs**: Added command classes and handlers for student creation and login operations.
- **appsettings.json**: Added JWT configuration settings.

aabdoo23 committed:
Bismillah

aabdoo23 committed:
first commit, entities not done yet

Add new entities, validators, and project references

Updated SchedsForums.Application.csproj and SchedsForums.Domain.csproj to include new package and project references. Cleaned up ConfigureServices.cs by removing unused using directives. Removed IBaseEntity.cs and IBaseRepository.cs from SchedsForums.Domain. Added new entity classes (Admin, Major, Moderator, Student) and base classes (BaseEntity, BaseUser). Introduced new validators and services, including ConfigureValidators.cs and PasswordService.cs. Created SchedsForums.Interface project with new interfaces and project references.

aabdoo23 committed:
Add persistence services and refactor domain models for the user sign up, next step is the repository, commands and queries, and unit of work

- Added `SchedsForums.Persistence.Services` to `Program.cs` to include persistence services.
- Modified `Program.cs` to add persistence services to the service collection and build the application.
- Removed `launchSettings.json` from the project.
- Updated `.gitignore` to include `FodyWeavers.xsd` and `SchedsForums.Api/Properties/launchSettings.json`.
- Changed `SchedsForums.Api.csproj` to include a project reference to `SchedsForums.Persistence`.
- Updated `ConfigureServices.cs` to return `IServiceCollection` from the `Configure` method.
- Modified `BaseEntity.cs` to make `CreatedAt` and `UpdatedAt` properties settable.
- Refactored `BaseUser.cs` to remove redundant code and improve readability.
- Updated `Admin.cs`, `Moderator.cs`, and `Student.cs` to include parameterless constructors.
- Reformatted `SchedsForums.Domain.csproj` and removed redundant project references.
- Added a project reference to `SchedsForums.Interface` in `SchedsForums.Infrastructure.csproj`.
- Refactored `PasswordService.cs` to implement `IBaseService`.
- Added `ForumsDbContext` class to manage database context and entity configurations.
- Created initial migration files for database schema.
- Updated `SchedsForums.Persistence.csproj` to include a project reference to `SchedsForums.Domain`.
- Added `DbContextRegistrationService.cs` to register the database context with dependency injection.

aabdoo23 committed:
repositories implemented

aabdoo23 committed:
repos fully done

Refactor namespaces and add repository services

Updated Program.cs to include new using directives and changed the namespace to SchedsForums.Api. Added calls to AddPersistenceServices and RegisterRepositories for DI. Refactored ConfigureServices.cs and ConfigureValidators.cs to use SchedsForums.Domain.ServiceConfiguration namespace. Updated SchedsForums.Infrastructure.csproj to reference SchedsForums.Domain. Added RepositoryRegistration.cs for repository service registration. Introduced new repository classes for Admin, Major, Moderator, and Student entities.

aabdoo23 committed:
Fixed the CLEAN Arch referencing issue, could use some refactoring later to ensure that everything in the right place

Did some work in the insert command for student section, but not done yet

Refactor namespaces and update project references

- Updated `Program.cs` to register persistence services and repositories.
- Modified `SchedsForums.Api.csproj` to reference `SchedsForums.Infrastructure`.
- Added `MediatR` package and folders for commands and queries in `SchedsForums.Application.csproj`.
- Cleaned up `SchedsForums.Domain.csproj` by removing unnecessary references.
- Updated `BaseEntity.cs` to use `SchedsForums.Domain.Interfaces` namespace.
- Refactored `BaseUser.cs` to `SchedsForums.Domain.Entities.Common` and simplified password handling.
- Removed unused using directives in `Major.cs`.
- Simplified exception message in `EntityNotFoundException.cs`.
- Updated repositories to use `SchedsForums.Domain.Interfaces` namespace.
- Updated `RepositoryRegistration.cs` to use `SchedsForums.Domain.Interfaces.Repositories` namespace.
- Cleaned up and updated `SchedsForums.Infrastructure.csproj` to include migrations folder and reference `SchedsForums.Domain`.
- Updated `PasswordService.cs` to use `SchedsForums.Domain.Interfaces` namespace.
- Updated `IMajorRepository.cs` to use `SchedsForums.Domain.Interfaces.Repositories` namespace.
- Updated `SchedsForums.Interface.csproj` to reference `SchedsForums.Domain`.
- Configured `BaseUser` entity with discriminator in `ForumsDbContext.cs`.
- Updated `SchedsForums.Persistence.csproj` to reference `SchedsForums.Domain`.
- Updated `SchedsForums.sln` to use new project GUIDs and removed references to `SchedsForums.Interface` and `SchedsForums.Persistence`.
- Added new DTOs, commands, and handlers for student creation.
- Added new interfaces for entities, services, and repositories.
- Added new `ForumsDbContext` in `SchedsForums.Infrastructure.Contexts`.
- Added `DbContextRegistrationService` in `SchedsForums.Persistence.Services`.

aabdoo23 committed:
CreateStudentCommand done (Sign up)

Refactor services, add MediatR, and update student creation

- Added `using` statements and updated `namespace` in `Program.cs`.
- Registered new services: `RegisterInfrastructureRepositories`, `RegisterInfrastructureServices`, `AddMediatRServices`.
- Added `MediatR` package and project references in `SchedsForums.Api.csproj`.
- Changed `CreateStudentCommand` return type to `StudentDTO`.
- Updated `CreateStudentCommandHandler` to use `FluentValidation` and `IPasswordService`.
- Modified `CreateStudentDTO` to inherit from `StudentDTO`.
- Simplified constructors in `BaseEntity` and `BaseUser`.
- Removed empty class body in `Major.cs` and `RepositoryRegistration` class.
- Added `Console.WriteLine` for connection string in `DbContextRegistrationService`.
- Refactored `PasswordService` to implement `IPasswordService`.
- Added `StudentsController` for student creation.
- Added initial migration files for database schema.
- Added `RegisterServices`, `CreateStudentValidator`, and `ServicesRegistration` classes.
- Added `IPasswordService` interface.

aabdoo23 committed:
final updates based on PR

aabdoo23 committed:
Refactored the structure and re-referenced the whole project.

Refactor project structure and update repository usage

Updated CreateStudentCommandHandler to use new interfaces from SchedsForums.Application. Modified SchedsForums.Application.csproj to reference SchedsForums.Domain. Simplified BaseEntity by removing IBaseEntity implementation. Refactored BaseUser, Admin, Moderator, and Student to remove base class constructor calls. Removed obsolete interfaces from SchedsForums.Domain.Interfaces. Reintroduced OnModelCreating in SchedsForumsDbContext. Replaced GenericRepository with BaseRepository. Updated StudentRepository to inherit from BaseRepository and use new IStudentRepository interface. Updated SchedsForums.Infrastructure.csproj to reference SchedsForums.Application. Updated HashingService to use IHashingService interface. Added new interfaces and BaseRepository class to SchedsForums.Application.Interfaces and SchedsForums.Infrastructure.Repositories.Common respectively.

aabdoo23 committed:
Add JWT auth, user management, and update migrations

- Removed old migration files and added new ones under `SchedsForums.Infrastructure.Migrations` namespace.
- Updated `Program.cs` to include `RegisterJwtAuth` for JWT setup.
- Updated `appsettings.json` with JWT `Issuer` and `Audience` settings.
- Added `UserName` property to `StudentRequestBaseDTO.cs`.
- Refactored `CreateStudentCommandHandler.cs` for constructor injection and updated to handle `UserName` and `PasswordHash`.
- Updated `CreateStudentValidator.cs` to validate `PasswordHash`.
- Renamed `Password` to `PasswordHash` and added `UserName` in `BaseUser.cs`.
- Refactored `StudentRepository.cs` for constructor injection.
- Updated `SchedsForums.Infrastructure.csproj` with JWT package references.
- Updated `ServicesRegistration.cs` to register JWT auth and services.
- Added `LoginController.cs` for handling login requests.
- Added `LoginCommand.cs` and `LoginCommandHandler.cs` for login logic.
- Added `LoginRequestDTO.cs` and `LoginResponseDTO.cs` for login data transfer.
- Added `IBaseUserRepository.cs` and `IAuthService.cs` for user retrieval and auth services.
- Added migration `20241022101035_update-user.cs` to update `Users` table.
- Re-added and updated initial migration files and `ForumsDbContextModelSnapshot.cs`.

MostafaHgz commented on SchedsForums.Api/Controllers/LoginController.cs:
```diff
@@ -0,0 +1,39 @@
+        [HttpPost]
+        public async Task<IActionResult> Login([FromBody] LoginCommand command)
+        {
+            if (command == null)
```
if command is null it should never get in here

MostafaHgz commented on SchedsForums.Api/Controllers/LoginController.cs:
```diff
@@ -0,0 +1,39 @@
+        {
+            if (command == null)
+                return BadRequest("Invalid login request");
+            try
```
will you use try catch in every action? 
use exception filter

MostafaHgz commented on SchedsForums.Api/Controllers/LoginController.cs:
```diff
@@ -0,0 +1,39 @@
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Login([FromBody] LoginCommand command)
```
LoginAsync

MostafaHgz commented on SchedsForums.Api/Controllers/LoginController.cs:
```diff
@@ -0,0 +1,39 @@
+{
+    [ApiController]
+    [Route("api/[controller]")]
+    public class LoginController : ControllerBase
```
This should be auth controller and it should have both login and signup

MostafaHgz commented on SchedsForums.Application/BaseDTOs/StudentRequestBaseDTO.cs:
```diff
@@ -0,0 +1,9 @@
+﻿namespace SchedsForums.Application.BaseDTOs
+{
+    public class StudentRequestBaseDTO
```
This should be UserSignUpRequestDTO

MostafaHgz commented on SchedsForums.Application/Commands/Students/Create/CreateStudentCommand.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace SchedsForums.Application.Commands.Students.Create
+{
+    public class CreateStudentCommand : StudentRequestBaseDTO, IRequest<StudentRequestBaseDTO>
```
it should be only one dto which is UserSignupRequestDTO

aabdoo23 replied:
Agreed

MostafaHgz commented on SchedsForums.Application/BaseDTOs/StudentRequestBaseDTO.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public class StudentRequestBaseDTO
+    {
+        public string Name { get; set; }
```
What is the difference between Name and UserName?

aabdoo23 replied:
Full name and username, username can be used for sign in and it's a unique handle, will add the sign in using username part

MostafaHgz commented on SchedsForums.Application/Commands/Students/Create/CreateStudentCommandHandler.cs:
```diff
@@ -0,0 +1,43 @@
+
+namespace SchedsForums.Application.Commands.Students.Create
+{
+    public class CreateStudentCommandHandler(IStudentRepository studentRepository, IHashingService passwordService, IValidator<CreateStudentCommand> validator) : IRequestHandler<CreateStudentCommand, StudentRequestBaseDTO>
```
add a line between every injected service

aabdoo23 replied:
okay

MostafaHgz commented on SchedsForums.Application/Commands/Students/Create/CreateStudentCommandHandler.cs:
```diff
@@ -0,0 +1,43 @@
+
+        public async Task<StudentRequestBaseDTO> Handle(CreateStudentCommand request, CancellationToken cancellationToken)
+        {
+            var validationResult = await _validator.ValidateAsync(request,cancellationToken);
```
didn't we discuss the idea that this is not the correct way to do it?

MostafaHgz commented on SchedsForums.Application/Commands/Students/Create/CreateStudentCommandHandler.cs:
```diff
@@ -0,0 +1,43 @@
+            };
+            await _studentRepository.InsertAsync(student);
+
+            return new StudentRequestBaseDTO
```
how is the response called Request?

MostafaHgz commented on SchedsForums.Application/Commands/Students/Create/CreateStudentDTO.cs:
```diff
@@ -0,0 +1,11 @@
+
+namespace SchedsForums.Application.Commands.Students.Create
+{
+    public class CreateStudentDTO : StudentRequestBaseDTO
```
what is the reason behind this?

MostafaHgz commented on SchedsForums.Application/Commands/Students/Login/DTOs/LoginRequestDTO.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace SchedsForums.Application.Commands.Students.Login.DTOs
+{
+    public class LoginRequestDTO
```
how is the login dto has no fields?

aabdoo23 replied:
I created it and the command but used the command only sorry

MostafaHgz commented on SchedsForums.Application/Commands/Students/Login/LoginCommandHandler.cs:
```diff
@@ -0,0 +1,32 @@
+
+        public LoginCommandHandler(IBaseUserRepository userRepository, IAuthService authService)
+        {
+            _userRepository = userRepository;
```
add null check

MostafaHgz commented on SchedsForums.Application/Commands/Students/Login/LoginCommandHandler.cs:
```diff
@@ -0,0 +1,32 @@
+
+            if (user == null || !_authService.VerifyPassword(user, request.Password))
+            {
+                throw new UnauthorizedAccessException("Invalid credentials");
```
this should be thrown inside the verifyPassword

MostafaHgz commented on SchedsForums.Application/Interfaces/Common/IBaseService.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace SchedsForums.Application.Interfaces.Common
+{
+    public interface IBaseService
```
this should be deleted

MostafaHgz commented on SchedsForums.Application/Interfaces/Repositories/IStudentRepository.cs:
```diff
@@ -0,0 +1,9 @@
+﻿using SchedsForums.Application.Interfaces.Common;
+using SchedsForums.Domain.Entities.Users;
+
+namespace SchedsForums.Application.Interfaces.Repositories
```
if it's empty then it does nothing, remove it

MostafaHgz commented on SchedsForums.Application/Interfaces/Services/IAuthService.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public interface IAuthService
+    {
+        bool VerifyPassword(BaseUser user, string password);
```
this should only be used in the login command so just make it there, no need to put it in a separate service

aabdoo23 replied:
i dont agree very much, i think its a service related to authentication, so in case there are other auth services they should be together in the same file?

MostafaHgz commented on SchedsForums.Application/Interfaces/Services/IAuthService.cs:
```diff
@@ -0,0 +1,10 @@
+    public interface IAuthService
+    {
+        bool VerifyPassword(BaseUser user, string password);
+        string GenerateToken(BaseUser user);
```
token handling should be in its own JWT service

MostafaHgz commented on SchedsForums.Application/RegisterServices.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace SchedsForums.Application
+{
+    public static class RegisterServices
```
rename to Application (or API) DIExtensions

MostafaHgz commented on SchedsForums.Application/RegisterServices.cs:
```diff
@@ -0,0 +1,23 @@
+{
+    public static class RegisterServices
+    {
+        public static IServiceCollection RegisterMediatRServices(this IServiceCollection services)
```
rename to AddMediatRService

MostafaHgz commented on SchedsForums.Application/RegisterServices.cs:
```diff
@@ -0,0 +1,23 @@
+
+        public static IServiceCollection RegisterValidationServices(this IServiceCollection services)
+        {
+            services.AddValidatorsFromAssemblyContaining<CreateStudentValidator>();
```
this way you'll need to add the validator each time you add a new one and when you have 100 endpoints this method will be a mess. check another way to add all validators

MostafaHgz commented on SchedsForums.Application/RegisterServices.cs:
```diff
@@ -0,0 +1,23 @@
+            services.AddValidatorsFromAssemblyContaining<CreateStudentValidator>();
+            return services;
+        }
+
```
remove empty space

MostafaHgz commented on SchedsForums.Domain/Entities/Common/BaseEntity.cs:
```diff
@@ -0,0 +1,15 @@
+        public DateTime CreatedAt { get; set; }
+        public DateTime UpdatedAt { get; set; }
+        public BaseEntity()
+        {
```
Id field here should be initialized by efcore itself. no need for this. 
created at and updated at should not be initialized in the constructor.

MostafaHgz commented on SchedsForums.Domain/Entities/Users/Admin.cs:
```diff
@@ -0,0 +1,8 @@
+﻿using SchedsForums.Domain.Entities.Common;
+
+namespace SchedsForums.Domain.Entities.Users
```
if it has nothing then it does nothing remove it

MostafaHgz commented on SchedsForums.Domain/Entities/Users/Moderator.cs:
```diff
@@ -0,0 +1,8 @@
+﻿using SchedsForums.Domain.Entities.Common;
+
+namespace SchedsForums.Domain.Entities.Users
+{
```
remove

MostafaHgz commented on SchedsForums.Infrastructure/Services/AuthService.cs:
```diff
@@ -0,0 +1,46 @@
+        private readonly IConfiguration _config;
+        public AuthService(IConfiguration configuration)
+        {
+            _config = configuration;
```
add null check

MostafaHgz commented on SchedsForums.Infrastructure/Services/AuthService.cs:
```diff
@@ -0,0 +1,46 @@
+        {
+            _config = configuration;
+        }
+        public string GenerateToken(BaseUser user)
```
token handling should be in its own service

MostafaHgz commented on SchedsForums.Infrastructure/Services/AuthService.cs:
```diff
@@ -0,0 +1,46 @@
+        }
+        public string GenerateToken(BaseUser user)
+        {
+            var key = Environment.GetEnvironmentVariable("JWT_KEY") ?? throw new NullReferenceException("Can't find JWT Key in Env Variables.");
```
this way you'll get the JWT_KEY from the environment with each login request. you only need to get it once. do it in the DI itself
also if you want to go the extra mile, check how to do it with IOptions

aabdoo23 replied:
ok

MostafaHgz commented on SchedsForums.Infrastructure/Services/AuthService.cs:
```diff
@@ -0,0 +1,46 @@
+            var claims = new[]
+            {
+                new Claim(ClaimTypes.NameIdentifier, user.Id),
+                new Claim(ClaimTypes.Name, user.UserName),
```
Username is a single word

MostafaHgz commented on SchedsForums.Infrastructure/ServicesRegistration.cs:
```diff
@@ -0,0 +1,62 @@
+
+namespace SchedsForums.Infrastructure
+{
+    public static class ServicesRegistration
```
don't use registration. rename this and the name should show that it contains services extension methods

MostafaHgz commented on SchedsForums.Infrastructure/ServicesRegistration.cs:
```diff
@@ -0,0 +1,62 @@
+{
+    public static class ServicesRegistration
+    {
+        public static IServiceCollection RegisterDbContext(this IServiceCollection services)
```
use Add instead of Register

MostafaHgz commented on SchedsForums.Infrastructure/ServicesRegistration.cs:
```diff
@@ -0,0 +1,62 @@
+            services.AddScoped<IBaseUserRepository, BaseUserRepository>();
+            return services;
+        }
+        public static IServiceCollection RegisterInfrastructureServices(this IServiceCollection services)
```
add a line between each method (do this across the whole solution

MostafaHgz commented on SchedsForums.Interface/IBaseEntity.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface IBaseEntity
+    {
+        string Id { get; }
```
you should separate the IBaseEntity and IAuditableEntity

MostafaHgz commented on SchedsForums.Interface/IBaseService.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace SchedsForums.Interface
```
if it's empty remove it

MostafaHgz replied:
remove it tany

MostafaHgz commented on SchedsForums.Interface/IGenericRepository.cs:
```diff
@@ -0,0 +1,12 @@
+        Task<T> InsertAsync(T entity);
+        Task<T> UpdateAsync(T entity);
+        Task<T> DeleteAsync(string id);
+
```
remove extra line

MostafaHgz commented on SchedsForums.Interface/IGenericRepository.cs:
```diff
@@ -0,0 +1,12 @@
+﻿namespace SchedsForums.Interface
+{
+    public interface IGenericRepository<T> where T : IBaseEntity
```
why do we have this file twice?

MostafaHgz commented on SchedsForums.Interface/Repositories/IAdminRepository.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace SchedsForums.Interface.Repositories
+{
+    public interface IAdminRepository: IGenericRepository<Admin>
```
if it's empty remove it

MostafaHgz replied:
again

MostafaHgz commented on SchedsForums.Interface/Repositories/IMajorRepository.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace SchedsForums.Interface.Repositories
+{
+    public interface IMajorRepository : IGenericRepository<Major>
```
if it's empty remove it

MostafaHgz replied:
again

MostafaHgz commented on SchedsForums.Interface/Repositories/IModeratorRepository.cs:
```diff
@@ -0,0 +1,8 @@
+
+namespace SchedsForums.Interface.Repositories
+{
+    public interface IModeratorRepository:IGenericRepository<Moderator>
```
if it's empty remove it

MostafaHgz replied:
also all of these are not in the scope of the task

MostafaHgz replied:
again

MostafaHgz commented on SchedsForums.Persistence/Contexts/ForumsDbContext.cs:
```diff
@@ -0,0 +1,32 @@
+        public DbSet<Student> Students { get; set; }
+        public DbSet<Moderator> Moderators { get; set; }
+
+
```
remove extra lines

MostafaHgz commented on SchedsForums.Persistence/Contexts/ForumsDbContext.cs:
```diff
@@ -0,0 +1,32 @@
+        }
+
+        public DbSet<BaseUser> Users { get; set; }
+        public DbSet<Admin> Admins { get; set; }
```
only user and student are in the scope of this story

MostafaHgz commented on SchedsForums.Persistence/Services/DbContextRegistrationService.cs:
```diff
@@ -0,0 +1,20 @@
+            var connectionString = Environment.GetEnvironmentVariable("ConnectionString");
+            services.AddDbContext<ForumsDbContext>(options =>
+                options.UseNpgsql((connectionString)
+                //, b => b.MigrationsAssembly("SchedsForums.Persistence")
```
what is this?

aabdoo23 committed:
PR 3 Revised

#3

Refactor controllers and add JWT authentication

Consolidated `LoginController` and `StudentController` into `AuthController`. Updated `Program.cs` for custom exception handling and service registration. Introduced `CustomExceptionFilter` for global exception handling. Added `ApplicationDIExtensions` for MediatR and FluentValidation DI. Implemented new authentication and user management functionalities with `LoginCommandHandler`, `UserSignUpCommandHandler`, and related DTOs. Updated `BaseEntity` and `BaseUser` for new structure. Removed obsolete files and methods. Added JWT authentication setup with `ServiceCollectionExtensions`, `JWTOptions`, and `JWTService`.

MostafaHgz commented on SchedsForums.Api/Controllers/AuthController.cs:
```diff
@@ -0,0 +1,34 @@
+            return Ok(result);
+
+        }
+        [HttpPost("SignUp")]
```
add line

MostafaHgz commented on SchedsForums.Api/Filters/CustomExceptionFilter.cs:
```diff
@@ -0,0 +1,29 @@
+            {
+                context.Result = new UnauthorizedObjectResult("Invalid credentials");
+                context.ExceptionHandled = true;
+            }
```
return inside the if and remove else

MostafaHgz replied:
refactor this

aabdoo23 replied:
7lw kda?

MostafaHgz commented on SchedsForums.Api/SchedsForums.Api.csproj:
```diff
@@ -7,11 +7,23 @@
   <ItemGroup>
-    <Folder Include="Controllers\" />
+    <ProjectReference Include="..\SchedsForums.Application\SchedsForums.Application.csproj" />
+    <ProjectReference Include="..\SchedsForums.Infrastructure\SchedsForums.Infrastructure.csproj" />
```
api should not depend on infrastructure

aabdoo23 replied:
tl3 ynf3

MostafaHgz commented on SchedsForums.Api/appsettings.json:
```diff
@@ -5,5 +5,9 @@
       "Microsoft.AspNetCore": "Warning"
     }
   },
+  "JWT": {
```
this move this to appSettings.Development.json

MostafaHgz commented on SchedsForums.Application/ApplicationDIExtensions.cs:
```diff
@@ -0,0 +1,27 @@
+            return services;
+        }
+
+        public static IServiceCollection AddValidationServices(this IServiceCollection services)
```
try 
`services.AddMediatR(configuration => configuration.RegisterServicesFromAssembly(Assembly.GetExecutingAssembly()));` and remove addValidations

MostafaHgz commented on SchedsForums.Application/Commands/BaseUser/Login/LoginCommandHandler.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace SchedsForums.Application.Commands.BaseUser.Login
+{
+    public class LoginCommandHandler(IBaseUserRepository userRepository, IAuthService authService, IJWTService jwtService) : IRequestHandler<LoginRequestDTO, LoginResponseDTO>
```
divide into multiple lines

MostafaHgz commented on SchedsForums.Application/Commands/BaseUser/Login/LoginCommandHandler.cs:
```diff
@@ -0,0 +1,24 @@
+        {
+            var user = await _userRepository.GetBaseUserByUserNameAsync(request.Username);
+
+            _authService.VerifyPassword(user, request.Password);
```
rename to password service

MostafaHgz commented on SchedsForums.Application/Commands/BaseUser/SignUp/DTOs/UserSignUpRequestDTO.cs:
```diff
@@ -0,0 +1,12 @@
+
+namespace SchedsForums.Application.Commands.BaseUser.SignUp.DTOs
+{
+    public class UserSignUpRequestDTO : IRequest<UserSignUpResponseDTO>
```
rename to command

MostafaHgz commented on SchedsForums.Application/Commands/BaseUser/Login/DTOs/LoginRequestDTO.cs:
```diff
@@ -0,0 +1,10 @@
+
+namespace SchedsForums.Application.Commands.BaseUser.Login.DTOs
+{
+    public class LoginRequestDTO : IRequest<LoginResponseDTO>
```
rename to command

MostafaHgz commented on SchedsForums.Application/Commands/BaseUser/SignUp/UserSignUpCommandHandler.cs:
```diff
@@ -0,0 +1,38 @@
+
+namespace SchedsForums.Application.Commands.BaseUser.SignUp
+{
+    public class UserSignUpCommandHandler(
```
rename to student

MostafaHgz commented on SchedsForums.Application/Commands/BaseUser/SignUp/UserSignUpCommandHandler.cs:
```diff
@@ -0,0 +1,38 @@
+        IHashingService passwordService)
+        : IRequestHandler<UserSignUpRequestDTO, UserSignUpResponseDTO>
+    {
+        private readonly IBaseRepository<Student> _studentRepository = studentRepository;
```
add null check

MostafaHgz commented on SchedsForums.Domain/Entities/Common/BaseEntity.cs:
```diff
@@ -0,0 +1,10 @@
+        public string Id { get; protected set; } = Guid.NewGuid().ToString();
+        public DateTime CreatedAt { get; set; }
+        public DateTime UpdatedAt { get; set; }
+        
```
remove extra line

MostafaHgz commented on SchedsForums.Domain/Entities/Common/BaseEntity.cs:
```diff
@@ -0,0 +1,10 @@
+{
+    public class BaseEntity
+    {
+        public string Id { get; protected set; } = Guid.NewGuid().ToString();
```
check this

aabdoo23 replied:
Guid works

MostafaHgz commented on SchedsForums.Infrastructure/Contexts/SchedsForumsDbContext.cs:
```diff
@@ -0,0 +1,45 @@
+
+        private void SetTimestamps()
+        {
+            var entries = ChangeTracker.Entries<BaseEntity>();
```
filter on IAuditable

MostafaHgz commented on SchedsForums.Infrastructure/Options/JWTOptions.cs:
```diff
@@ -0,0 +1,10 @@
+        public string Key { get; set; } = string.Empty;
+        public string Issuer { get; set; } = string.Empty;
+        public string Audience { get; set; } = string.Empty;
+        public int ExpirationMinutes { get; set; } = 15;
```
use this in a constant

MostafaHgz commented on SchedsForums.Infrastructure/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,79 @@
+        {
+            services.Configure<JwtOptions>(options =>
+            {
+                configuration.GetSection("Jwt").Bind(options);
```
add null checks on other options

MostafaHgz commented on SchedsForums.Infrastructure/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,79 @@
+                    ValidateAudience = true,
+                    ValidateLifetime = true,
+                    ValidateIssuerSigningKey = true,
+                    ValidIssuer = configuration["Jwt:Issuer"],
```
use jwtOptions

MostafaHgz commented on SchedsForums.Infrastructure/Services/AuthService.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace SchedsForums.Infrastructure.Services
+{
+    public class AuthService : IAuthService
```
password

MostafaHgz commented on SchedsForums.Persistence/Contexts/ForumsDbContext.cs:
```diff
@@ -0,0 +1,24 @@
+        public ForumsDbContext(DbContextOptions<ForumsDbContext> options) : base(options)
+        {
+        }
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
```
add extra line before

MostafaHgz commented on SchedsForums.Persistence/Contexts/ForumsDbContext.cs:
```diff
@@ -0,0 +1,24 @@
+        }
+        protected override void OnModelCreating(ModelBuilder modelBuilder)
+        {
+            modelBuilder.Entity<BaseUser>()
```
get config from assembly

aabdoo23 replied:
this deleted

MostafaHgz commented on SchedsForums.Persistence/SchedsForums.Persistence.csproj:
```diff
@@ -1,9 +1,31 @@
-    <ImplicitUsings>enable</ImplicitUsings>
-    <Nullable>enable</Nullable>
-  </PropertyGroup>
+	<PropertyGroup>
```
remove

aabdoo23 committed:
#3
PR Revised

Refactor auth and service registration; remove old DTOs

Refactored AuthController to use new command classes for login and sign-up. Updated service registration in Program.cs to use combined methods. Added FluentValidation package and JWT settings. Removed obsolete DTOs, handlers, and interfaces. Introduced new command handlers, DTOs, and validators for login and sign-up. Changed entity ID types to Guid. Removed redundant projects and files.

MostafaHgz commented on SchedsForums.Api/Controllers/AuthController.cs:
```diff
@@ -0,0 +1,35 @@
+
+        public AuthController(IMediator mediator)
+        {
+            _mediator = mediator;
```
add null check

MostafaHgz commented on SchedsForums.Api/Filters/CustomExceptionFilter.cs:
```diff
@@ -0,0 +1,51 @@
+{
+    public class CustomExceptionFilter : IExceptionFilter
+    {
+        public void OnException(ExceptionContext context)
```
todo enhance this

MostafaHgz commented on SchedsForums.Infrastructure/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,85 @@
+        }
+
+        public static IServiceCollection AddJwtAuthentication(
+            this IServiceCollection services, 
```
afsfsd

aabdoo23 committed:
#3

PR DONE >:(

Refactor DbContext, add JWT config, and update migrations

Refactored SchedsForumsDbContext to use a constructor. Added null
check for _mediator in AuthController. Configured JWT options in
Program.cs. Updated ServiceCollectionExtensions to improve JWT
configuration. Removed old migration files and added new migration
for Users table. Added TODO comment in CustomExceptionFilter.
